### DISCUSSION TITLE
Project Structure with Helper functions

### DISCUSSION THREAD

**Discussion by jask06** [NationalSecurityAgency/ghidra#6572] [Q&A] [answered]
Hello! I have a question regarding program structure when using the headless analyzer to run custom Ghidra Scripts.

I am currently looking at a project with the following structure:

```bash
project-home/
├── modules-directory
│   ├── module.java
│   └── module2.java
└── scripts-directory
    ├── Script.java
    └── Script2.java
```

At this time, when running `ghidra/support/analyzeHeadless` the `scripts-directory` is defined as the -scriptPath. I am looking to introduce modules/helper functions that can be used by `ScriptX.java` files in the `modules-directory`. Currently analysis works great, when nothing references `modules-directory`.

When I do introduce references to `modules-directory`, the resulting scan returns `cannot find symbol errors`, such as below:

```bash
Script.java:38: error: cannot find symbol
          Module.moduleMethod(<params>);
```

Is this possible to use modules/helper functions with the structure above, because I am currently not having any success. I have expanded the scope of -scriptPath to include both the `scripts-directory` and `modules-directory`. I have also tried adding the `modules-directory` path to the `-cp` flag in `launch.sh` to see if the run would pick up the modules. I also currently have my UserPluginPath set to `project-home`, but I have tried adjusting it to see if that makes any difference, but I also suspect it wouldn't pick up `.java` files anyway, is that correct? Finally, I have looked into compiling the module into a `.jar` and placing it into the GHIDRA_INSTALL_DIR without success.

Could anyone confirm if the keeping the structure above can work or is moving the `modules-directory` underneath the `scripts-directory` the preferred/prescribed way of achieving this functionality (which I have confirmed to work).  

Thank you in advance for your help!

**Comment by dragonmacher** [comment, marked as answer]
Things have changed so much over the years that I cannot remember the current preferred way to get something like this working.  I believe now that we recommend all scripting use the Bundle system.  This is done for free under the hood for basic script creation.  For more advanced usage, such as sharing code across modules, or distributing scripts, you should be using bundles. 

That being said, at the most fundamental level, how we compile and run scripts should all based on the classpath of Ghidra.  You should be able to access any code on Ghidra's classpath from within your script.   Assuming your module is on Ghidra's classpath, then your script should have access to the helper functions at runtime.   I still think creating the correct bundle structure, with all scripts and helper functions in that bundle (or in multiple bundles, with dependencies specified), and then building that bundle(s) before running the headless analyzer is probably the preferred mechanism.

**Reply by jask06** [reply]
Thank you very much for the response! I'll look further into the Bundle system and go from there!

Much appreciated!
